File: backend/internal/repository/user_repository.go
package repository

import (
	"context"
	"database/sql"

	"github.com/jmoiron/sqlx"

	"topsus-krs/internal/domain"
)

// UserRepository mendefinisikan kontrak operasi database untuk entitas User
type UserRepository interface {
	Create(ctx context.Context, user *domain.User) error
	FindByEmail(ctx context.Context, email string) (*domain.User, error)
	FindByID(ctx context.Context, id uint64) (*domain.User, error)
	Update(ctx context.Context, user *domain.User) error
}

type userRepository struct {
	db *sqlx.DB
}

func NewUserRepository(db *sqlx.DB) UserRepository {
	return &userRepository{db: db}
}

func (r *userRepository) Create(ctx context.Context, user *domain.User) error {
	query := `
		INSERT INTO users (name, email, password, role)
		VALUES (:name, :email, :password, :role)
	`
	result, err := r.db.NamedExecContext(ctx, query, user)
	if err != nil {
		return err
	}
	id, err := result.LastInsertId()
	if err != nil {
		return err
	}
	user.ID = uint64(id)
	return nil
}

func (r *userRepository) FindByEmail(ctx context.Context, email string) (*domain.User, error) {
	var user domain.User
	query := `SELECT * FROM users WHERE email = ? LIMIT 1`
	err := r.db.GetContext(ctx, &user, query, email)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &user, err
}

func (r *userRepository) FindByID(ctx context.Context, id uint64) (*domain.User, error) {
	var user domain.User
	query := `SELECT * FROM users WHERE id = ? LIMIT 1`
	err := r.db.GetContext(ctx, &user, query, id)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &user, err
}

func (r *userRepository) Update(ctx context.Context, user *domain.User) error {
	query := `
		UPDATE users
		SET name = :name, email = :email, password = :password, updated_at = NOW()
		WHERE id = :id
	`
	_, err := r.db.NamedExecContext(ctx, query, user)
	return err
}

File: backend-refactor-python/app/repositories/user_repository.py
from datetime import datetime
from typing import Optional

from sqlalchemy import text
from sqlalchemy.orm import Session


class UserRecord:
    __slots__ = ("id", "name", "email", "password", "role", "created_at", "updated_at")

    def __init__(self, row):
        self.id         = row.id
        self.name       = row.name
        self.email      = row.email
        self.password   = row.password
        self.role       = row.role
        self.created_at = row.created_at
        self.updated_at = row.updated_at


class UserRepository:
    def __init__(self, db: Session):
        self.db = db

    def find_by_email(self, email: str) -> Optional[UserRecord]:
        row = self.db.execute(
            text("SELECT * FROM users WHERE email = :email LIMIT 1"),
            {"email": email},
        ).fetchone()
        return UserRecord(row) if row else None

    def find_by_id(self, user_id: int) -> Optional[UserRecord]:
        row = self.db.execute(
            text("SELECT * FROM users WHERE id = :id LIMIT 1"),
            {"id": user_id},
        ).fetchone()
        return UserRecord(row) if row else None

    def create(self, name: str, email: str, password: str, role: str) -> UserRecord:
        result = self.db.execute(
            text(
                "INSERT INTO users (name, email, password, role) "
                "VALUES (:name, :email, :password, :role)"
            ),
            {"name": name, "email": email, "password": password, "role": role},
        )
        self.db.commit()
        new_id = result.lastrowid
        return self.find_by_id(new_id)

    def update_name(self, user_id: int, name: str) -> None:
        self.db.execute(
            text("UPDATE users SET name = :name, updated_at = NOW() WHERE id = :id"),
            {"name": name, "id": user_id},
        )
        self.db.commit()

    def update_password(self, user_id: int, hashed: str) -> None:
        self.db.execute(
            text("UPDATE users SET password = :pw, updated_at = NOW() WHERE id = :id"),
            {"pw": hashed, "id": user_id},
        )
        self.db.commit()
